### PR TITLE
interface generation (a different take)

### PR DESCRIPTION
**Pull request by jacogr**
Probably more for discussion than merge. Took the approach -
- ES2015 interface definitions
- Generator that builds from this interface via `npm run build`
- Output as both index.js & index.json (latter in the same original format)
- Tests verify (commented out) that all original methods are passed-through
- Added a release script that can push the build to npm & update GitHub with latest index.js/json
- Some extra flags already added, e.g. db_\* has deprecated flag
- generated index.json has `"methods": [{ "name": "eth_fullName" }...]` (as original)
- generated index.js has an object structure `{ eth: { fullName: {...},,, }], db: {..}, ...}`

Basically an approach that can evolve and is more flexible. Some things I would like to use this as the basis for
- Standardised type definitions
- Adding client types, i.e. Geth won't support ethcore_, Parity doesn't do db_
- more to come...

### CODE REVIEW

**Comment by tomusdrw** [comment]
:+1: :+1: as always! Great work. LGTM!

**Comment by goldylucks** [comment]
very cool!

@jacogr is it ready for merge?

**Comment by jacogr** [comment]
Ready for merge, there is one caveat though -

I have not added the generated `index.js` & `index.json` files in here. I would expect travis to generate these, push to GitHub and then release to npm with a patch version update. For that travis needs the `NPM_[USERNAME,PASSWORD,EMAIL]` either encoded in `travis.yml` or in the env config. (used in `scripts/release.sh`)

(The whole publish part hasn't been tested in this branch, but similar to what I use elsewhere, so hopefully once configured...)
